skunkworks_crow/src/main/java/org/odk/share/tasks/UploadJob.java

```diff
@@ -89,34 +95,62 @@ protected Result onRunJob(@NonNull Params params) {
 
         initJob(params);
 
-        rxEventBus.post(uploadInstances());
-
         return null;
     }
 
     private void initJob(Params params) {
-        instancesToSend = ArrayUtils.toObject(params.getExtras().getLongArray(INSTANCES));
-        port = params.getExtras().getInt(PORT, -1);
-        mode = params.getExtras().getInt(MODE, ApplicationConstants.ASK_REVIEW_MODE);
         sbResult = new StringBuilder();
+        boolean isBluetooth = params.getExtras().getBoolean("isBluetooth", true);
+        mode = params.getExtras().getInt(MODE, ApplicationConstants.ASK_REVIEW_MODE);
+        instancesToSend = ArrayUtils.toObject(params.getExtras().getLongArray(INSTANCES));
+        if (!isBluetooth) {
+            port = params.getExtras().getInt(PORT, -1);
+        }
+
+        setupDataStreamsAndRun(isBluetooth);
     }
 
-    private UploadEvent uploadInstances() {
+    private void setupDataStreamsAndRun(boolean isBluetooth) {
         try {
             Timber.d("Waiting for receiver");
+            if (isBluetooth) {
+                BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
+                BluetoothServerSocket bluetoothServerSocket = bluetoothAdapter.listenUsingInsecureRfcommWithServiceRecord(TAG, SPP_UUID);
+                BluetoothSocket bluetoothSocket = bluetoothServerSocket.accept();
 
-            serverSocket = new ServerSocket(port);
-            socket = serverSocket.accept();
-            dos = new DataOutputStream(socket.getOutputStream());
-            dis = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+                if (bluetoothSocket.isConnected()) {
+                    rxEventBus.post(new BluetoothEvent(BluetoothEvent.Status.CONNECTED));
+                }
 
+                dos = new DataOutputStream(bluetoothSocket.getOutputStream());
+                dis = new DataInputStream(bluetoothSocket.getInputStream());
+            } else {
+                serverSocket = new ServerSocket(port);
+                socket = serverSocket.accept();
+                dos = new DataOutputStream(socket.getOutputStream());
+                dis = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+            }
+
+            rxEventBus.post(uploadInstances());
+        } catch (IOException e) {
+            Timber.e(e);
+        }
+    }
+
+    private UploadEvent uploadInstances() {
+        try {
             // show dialog and connected
             Timber.d("Start Sending");
             processSelectedFiles(instancesToSend);
 
             // close connection
-            socket.close();
-            serverSocket.close();
+            if (socket != null) {
```

Comment:
When all the form transfer is done newly created `bluetoothSocket` also needs to closed for both sender and receiver other it may result in memory leaks.

Reply:
Yes, thanks for reminding. Fixed.